### PR TITLE
Ban println/printf via clj-kondo config

### LINKED ISSUE #27511: Add Kondo linter to disallow `println` (instead of `log/`)
You're always supposed to use stuff like `log/info` so we can configure it (e.g. disable it, especially in tests) -- but currently this is a rule that you just have to know (and PR reviewers have to catch it if you didn't know the rule). This is something we can easily have Kondo enforce for us. 

Not sure if this should only apply to `src` and the like and not apply to `test/`, or if there are good reasons to be doing `println` in tests (I think maybe we're currently doing it for some test runner init messages?) but we can always add`#{:clj-kondo/ignore ...}` in places where we have legitimate reasons to use it.

If we decide it should only apply to non-test directories we can create a namespace group in Kondo and only apply it to certain namespaces.

More ideas: maybe we should also make sure error messages inside `ex-info` and the like are i18n'ed; or messages inside `log/` statements higher than the `:debug` level are i18n'ed with `trs` (the site locale). Writing linters for these rules would be a little more complicated since we'd probably have to write a custom hook but it would be nice if Kondo could catch this rather than people having to know or PR reviewers having to look out for it

Related: #27250

### PR DESCRIPTION
**Pull request by markbastian**
Fixes #27511

### CODE REVIEW

File: .clj-kondo/config.edn [hunk 1]
```diff
@@ -83,7 +83,9 @@
     schema.core/either                                                           {:namespaces ["metabase.*"]}}}
 
   :discouraged-var
```
Comment (markbastian) [inline .clj-kondo/config.edn:85]: This is the primary change here. Adding bans on println and printf.

File: .clj-kondo/config.edn [hunk 2]
```diff
@@ -657,14 +659,19 @@
...
  {test-namespaces
   {:linters
-   {:inline-def        {:level :off}
-    :missing-docstring {:level :off}
-    :private-call      {:level :off}
+   {:inline-def                                    {:level :off}
+    :missing-docstring                             {:level :off}
+    :private-call                                  {:level :off}
 
     ;; custom linters
```
Comment (markbastian) [inline .clj-kondo/config.edn:673]: Note that I'd like to fine-grain this for metabase.cmd nses to not turn all the discouraged vars off, but kondo doesn't support this. I did put in a [kondo ticket](https://github.com/clj-kondo/clj-kondo/issues/1979) for this.

File: test/metabase/driver/common/parameters/dates_test.clj [hunk 3]
```diff
@@ -233,7 +233,7 @@
         (str frame n unit "-from-0" unit2)])
      (gen/tuple
       (gen/elements #{"next" "past"})
```
Comment (markbastian) [inline test/metabase/driver/common/parameters/dates_test.clj:235]: Fixed a couple other kondo warnings on the fly as well.

File: test/metabase/util_test.clj [hunk 4]
```diff
@@ -329,15 +329,15 @@
               (shuffle (range 30))))))
```
Comment (markbastian) [inline test/metabase/util_test.clj:330]: Other kondo fixes.

File: .clj-kondo/config.edn [hunk 5]
```diff
@@ -83,7 +83,9 @@
     schema.core/either                                                           {:namespaces ["metabase.*"]}}}
 
   :discouraged-var
-  {clojure.string/lower-case {:message "Use metabase.util/lower-case-en instead of clojure.string/lower-case"}
+  {clojure.core/println      {:message "Use clojure.tools.logging instead of clojure.core/println"}
+   clojure.core/printf       {:message "Use clojure.tools.logging instead of clojure.core/printf"}
```
Comment (escherize) [inline .clj-kondo/config.edn:87]: Is it worth adding `clojure.core/print` here?

Comment (markbastian) [reply .clj-kondo/config.edn:87]: Yeah, maybe prn as well.

Comment (markbastian) [reply .clj-kondo/config.edn:88]: Added print, prn, and pr to the config. Will fix next, but pushing this will prevent auto-merge.

Comment (markbastian) [reply .clj-kondo/config.edn:88]: Added all the prints....

File: .clj-kondo/config.edn [hunk 6]
```diff
@@ -92,14 +97,17 @@
 
   :unresolved-var
   {:exclude
-   [colorize.core]}
+   [colorize.core
```
Comment (markbastian) [inline .clj-kondo/config.edn:100]: Saw these were also shown as kondo issues, so added them. They were indeed required.

File: .clj-kondo/hooks/common.clj [hunk 7]
```diff
@@ -46,9 +46,10 @@
   ;; unused variables since Kondo was smart enough to realize that we sometimes were calling pure functions early in
   ;; the `do` and the value was getting thrown away. `print` tricks Kondo into thinking everything is used.
   [{{[_ & args] :children} :node}]
```
Comment (markbastian) [inline .clj-kondo/hooks/common.clj:48]: Is this (i.e. `(constantly nil)`) a legit substitute for `print` in these cases? I'm thinking so, but wanted a second opinion.

Comment (escherize) [reply .clj-kondo/hooks/common.clj:48]: I think that substitute works ok ~but maybe the log/info token would work better.~

**Review by escherize — APPROVED** [review]

**Review by qnkhuat — COMMENTED** [review]
If I'm not mistaken, during development and testing we config log to FATAL([source](https://github.com/metabase/metabase/blob/msb-ban-println-27511/test_config/log4j2-test.xml)).

And I find the println in drivers tests to be helpful, like `Creating new snowflake database %s...`.

Any reasons we want to get rid of those? if not I'd love to keep them as just print.

Also, should we ban `clojure.pprint/pprint` too?

File: .clj-kondo/hooks/common.clj (hunk 7 above)
Comment (tsmacdonald) [reply .clj-kondo/hooks/common.clj:48]: Putting `print` in was my hack from a couple weeks ago, and replacing it with `(constantly nil)` is definitely clearer and has my enthusiastic endorsement. If the Kondo check passes it's a legit substitute, I don't think there's anything subtle to consider.

Comment (tsmacdonald) [reply .clj-kondo/hooks/common.clj:48]: (that said, please update the comment too :) )

File: .clj-kondo/hooks/metabase/test/data.clj [hunk 8]
```diff
@@ -67,7 +68,7 @@
                ;; (this used to be a `do` form (which makes a lot more semantic sense), but that resulted in warnings
                ;; about unused values
                (:unqualified/from-dataset-defs-namespace :unqualified/unknown)
-               (list* (hooks/token-node 'print)
```
Comment (tsmacdonald) [inline .clj-kondo/hooks/metabase/test/data.clj:70]: :clap:

File: .clj-kondo/hooks/common.clj (hunk 7 above)
Comment (markbastian) [reply .clj-kondo/hooks/common.clj:48]: Done!

**Pull request merged**